### PR TITLE
Themes: Site picker for upload page

### PR DESCRIPTION
**Pull request by seear**
screen shot 2017-01-26 at 17 10 34

Show the site picker page on the /design/upload route when no site is selected.

This PR introduces a new middleware for adding the site picker to the layout without rendering, which is necessary for use with the single-tree-layouts used on the /design routes.

**To Test**
* Go to http://calypso.localhost:3000/design/upload
* **Expected:** The site picker should be displayed, selecting a site should take you to the upload page
* Also check an existing use of the site picker, for example http://calypso.localhost:3000/menus

### CODE REVIEW

**Comment by folletto** [comment]
Tested on Chrome, works as expected. Neat. 👍

File: client/my-sites/controller.js [hunk 1]
```diff
@@ -338,41 +369,45 @@ module.exports = {
...
+	 * without rendering the layout. For use with isomorphic routing
+	 * @see {@link https://github.com/Automattic/wp-calypso/blob/master/docs/isomorphic-routing.md }
+	 *
+	 * To show the site selector screen using traditional multi-tree
+	 * layout, use the sites() middleware above.
+	 *
+	 * @param {object} context -- Middleware context
+	 * @param {function} next -- Call next middleware in chain
+	 */
+	sitesWithoutRender( context, next ) {
```
Comment (ockham) [inline client/my-sites/controller.js]: My nomenclature was to prefix with `make`, i.e. `makeSites`, in this case. Cf [`makeNavigation`](https://github.com/Automattic/wp-calypso/blob/2725025cf4a3d808291acb3d312910b337f937f6/client/my-sites/controller.js#L303).

(Yay for the JSDoc!)

File: client/my-sites/controller.js [hunk 2]
```diff
@@ -154,13 +156,42 @@ function onSelectedSiteAvailable( context ) {
...
+function createSitesComponent( context ) {
+	const basePath = route.sectionify( context.path );
+	const path = context.prevPath ? route.sectionify( context.prevPath ) : '/stats';
+
+	// This path sets the URL to be visited once a site is selected
+	const sourcePath = ( basePath === '/sites' ) ? path : basePath;
+
+	analytics.pageView.record( basePath, sitesPageTitleForAnalytics );
+
+	return React.createElement( SitesComponent, {
```
Comment (ockham) [inline client/my-sites/controller.js]: You could use JSX here -- nevermind the filename suffix 😄

File: client/my-sites/controller.js [hunk 3]
```diff
@@ -33,12 +33,14 @@ import { setLayoutFocus } from 'state/ui/layout-focus/actions';
...
 import isDomainOnlySite from 'state/selectors/is-domain-only-site';
 import { domainManagementList } from 'my-sites/upgrades/paths';
+import SitesComponent from 'my-sites/sites';
 
 /**
  * Module vars
  */
 const user = userFactory();
 const sites = sitesFactory();
+const sitesPageTitleForAnalytics = 'Sites';
```
Comment (budzanowski) [inline client/my-sites/controller.js]: I love self describing code :+1:

**Comment by budzanowski** [comment]
Works nice, code looks nice! :shipit:

**Comment by budzanowski** [comment]
One thing. One opening `upload` i first see flash of `design`. 
flash
I don't know if this is related to this PR or not.

**Comment by seear** [comment]
> One thing. One opening upload i first see flash of design.

I believe this is because in the `development` env, the client initially thinks the user is logged-out, so we see a flash of the logged-out SSR'd /design page. I'll verify that this works ok after deployment.

File: client/my-sites/controller.js [hunk 4]
```diff
@@ -338,41 +363,45 @@ module.exports = {
...
-		analytics.pageView.record( basePath, analyticsPageTitle );
-
 		renderWithReduxStore(
-			React.createElement( SitesComponent, {
-				sites,
-				path: context.path,
-				sourcePath,
-				user,
-				getSiteSelectionHeaderText: context.getSiteSelectionHeaderText,
-				trackScrollPage: trackScrollPage.bind(
```
Comment (seear) [inline client/my-sites/controller.js]: the sites component actually does not take a `trackScrollPage` prop.

**Comment by seear** [comment]
> I'll verify that this works ok after deployment.

No content flash in production.

**Pull request merged**